kstreams/engine.py

```diff
@@ -133,18 +141,154 @@ async def send(
         return metadata
 
     async def start(self) -> None:
+        # Execute on_startup hooks
+        await execute_hooks(self._on_startup)
+
         # add the producer and streams to the Monitor
         self.monitor.add_producer(self._producer)
         self.monitor.add_streams(self._streams)
 
         await self.start_producer()
         await self.start_streams()
 
+        # Execute after_startup hooks
+        await execute_hooks(self._after_startup)
+
+    def on_startup(
+        self,
+        func: typing.Callable[[], typing.Any],
+    ) -> typing.Callable[[], typing.Any]:
+        """
+        A list of callables to run before the engine starts.
+        Handler are callables that do not take any arguments, and may be either
+        standard functions, or async functions.
+
+        Attributes:
+            func typing.Callable[[], typing.Any]: Func to callable before engine starts
+
+        !!! Example
+            ```python title="Engine before startup"
+
+            import kstreams
+
+            stream_engine = kstreams.create_engine(
+                title="my-stream-engine"
+            )
+
+            @stream_engine.on_startup
+            async def init_db() -> None:
+                print("Initializing Database Connections")
+                await init_db()
+
+
+            @stream_engine.on_startup
+            async def start_background_task() -> None:
+                print("Some background task")
+            ```
+        """
+        self._on_startup.append(func)
+        return func
+
+    def on_stop(
+        self,
+        func: typing.Callable[[], typing.Any],
+    ) -> typing.Callable[[], typing.Any]:
+        """
+        A list of callables to run before the engine stops.
+        Handler are callables that do not take any arguments, and may be either
+        standard functions, or async functions.
+
+        Attributes:
+            func typing.Callable[[], typing.Any]: Func to callable before engine stops
+
+        !!! Example
+            ```python title="Engine before stops"
+
+            import kstreams
+
+            stream_engine = kstreams.create_engine(
+                title="my-stream-engine"
+            )
+
+            @stream_engine.on_stop
+            async def close_db() -> None:
+                print("Closing Database Connections")
+                await db_close()
+            ```
+        """
+        self._on_stop.append(func)
+        return func
+
+    def after_startup(
+        self,
+        func: typing.Callable[[], typing.Any],
+    ) -> typing.Callable[[], typing.Any]:
+        """
+        A list of callables to run after the engine starts.
+        Handler are callables that do not take any arguments, and may be either
+        standard functions, or async functions.
+
+        Attributes:
+            func typing.Callable[[], typing.Any]: Func to callable after engine starts
```

Comment:
do you need the type in this comment?

Reply:
Yeah, it is kind of bad but it is needed otherwise the type is not shown in the docs :-(